server/cluster/unsafe_recovery_controller.go

```diff
@@ -121,452 +172,664 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
+			return rs.GetHardState().GetCommit() < os.GetHardState().GetCommit()
 		}
 	}
-	region.Peers = newPeerList
+	return false
 }
 
-type peerStorePair struct {
-	peer    *pdpb.PeerReport
-	storeID uint64
+const (
+	defaultBTreeDegree = 64
+)
+
+type regionTree struct {
+	regions map[uint64]*regionItem
+	tree    *btree.BTree
 }
 
-func getOverlapRanges(tree *btree.BTree, region *metapb.Region) []*metapb.Region {
-	var overlapRanges []*metapb.Region
-	tree.DescendLessOrEqual(regionItem{region}, func(item btree.Item) bool {
-		if bytes.Compare(item.(regionItem).region.StartKey, region.StartKey) < 0 && bytes.Compare(item.(regionItem).region.EndKey, region.StartKey) > 0 {
-			overlapRanges = append(overlapRanges, item.(regionItem).region)
+func newRegionTree() *regionTree {
+	return &regionTree{
+		regions: make(map[uint64]*regionItem),
+		tree:    btree.New(defaultBTreeDegree),
+	}
+}
+
+func (t *regionTree) contains(regionID uint64) bool {
+	_, ok := t.regions[regionID]
+	return ok
+}
+
+// getOverlaps gets the regions which are overlapped with the specified region range.
+func (t *regionTree) getOverlaps(item *regionItem) []*regionItem {
+	// note that find() gets the last item that is less or equal than the region.
+	// in the case: |_______a_______|_____b_____|___c___|
+	// new region is     |______d______|
+	// find() will return regionItem of region_a
+	// and both startKey of region_a and region_b are less than endKey of region_d,
+	// thus they are regarded as overlapped regions.
+	result := t.find(item)
+	if result == nil {
+		result = item
+	}
+
+	end := item.Region().GetEndKey()
+	var overlaps []*regionItem
+	t.tree.AscendGreaterOrEqual(result, func(i btree.Item) bool {
+		over := i.(*regionItem)
+		if len(end) > 0 && bytes.Compare(end, over.Region().GetStartKey()) <= 0 {
+			return false
 		}
+		overlaps = append(overlaps, over)
+		return true
+	})
+	return overlaps
+}
+
+// find is a helper function to find an item that contains the regions start key.
+func (t *regionTree) find(item *regionItem) *regionItem {
+	var result *regionItem
+	t.tree.DescendLessOrEqual(item, func(i btree.Item) bool {
+		result = i.(*regionItem)
 		return false
 	})
 
-	tree.AscendGreaterOrEqual(regionItem{region}, func(item btree.Item) bool {
-		if len(region.EndKey) != 0 && bytes.Compare(item.(regionItem).region.StartKey, region.EndKey) > 0 {
+	if result == nil || !result.Contains(item.Region().GetStartKey()) {
+		return nil
+	}
+
+	return result
+}
+
+// update updates the tree with the region.
+// It finds and deletes all the overlapped regions first, and then
+// insert the new region.
+func (t *regionTree) update(item *regionItem) bool {
+	overlaps := t.getOverlaps(item)
+
+	foundSelf := false
+	for _, old := range overlaps {
+		if item.Region().GetId() == old.Region().GetId() {
+			foundSelf = true
+			// not only check epoch for peers of the same region
+			if item.IsStale(old) {
+				return false
+			}
+		} else if item.IsEpochStale(old) {
```

Comment:
I am not sure this is what we really want. For example, imagine that there are multiple overlaps, the first overlap is newer (than the one is currently examining), and the second one is older, this code returns false after the first overlap is checked. I know that we not longer pursue "best effort", but ignoring the "newer" one but keeping the "older" one seems also problematic? Shouldn't we also remove the second overlap from the tree map as well?